astropy/wcs/src/wcslib_wrap.c

```diff
@@ -1517,14 +1691,35 @@ PyWcsprm_s2p(
     return NULL;
   }
 
-  if (PyArray_DIM(world, 1) < naxis) {
+  ncoord = (int)PyArray_DIM(world, 0);
+  nelem = (int)PyArray_DIM(world, 1);
+
+  if (nelem < naxis) {
     PyErr_Format(
       PyExc_RuntimeError,
       "Input array must be 2-dimensional, where the second dimension >= %d",
       naxis);
     goto exit;
   }
 
+  // Here we force a call to wcsset. Normally, WCSLIB will call wcsset automatically when
+  // calling wcsp2s, but we need to call it ourselves using PyWcsprm_cset so that we can
+  // catch cases where the units might change if e.g. they are not in SI to start with.
+  /* Force a call to wcsset here*/
+  if (self->preserve_units && PyWcsprm_cset(self, 1)) return NULL;
```

Comment:
What is the performance cost for this for the case of a single coordinate transformation (a pair of coordinates) and for a large array of coordinates?

Again, `{ return NULL; }`

Reply:
Fixed - I'll address performance in the main discussion below